File: frontend/types/sensors.ts
type SensorsType = {
    device_id: string
    name: string
    description?: string,
    id: number,
    created_at: string,
    updated_at: string
    timestamp: string
    value: number | string
    type: "sensors"
    unit: string
    details: Record<string, any>
    online: boolean
}

type UpdateSensorType = {
    "device_id": string,
    "name"?: string,
    "description"?: string,
    "id": number,
}

type SensorDataReadType = {
    device_id: string
    data: Record<string, any>
    timestamp: string
    value: number | string
    unit: string
    online: boolean
}

export type {SensorsType, UpdateSensorType, SensorDataReadType}

File: backend/schemas/sensors.py
from datetime import datetime
from typing import Any, Dict, Optional

from pydantic import BaseModel


class SensorMessage(BaseModel):
    device_id: str
    timestamp: str
    data: Dict[str, Any]
    value: Optional[float | str] = None
    unit: str
    online: Optional[bool] = None


class SensorBaseSchema(BaseModel):
    device_id: str
    name: str
    description: Optional[str]


class SensorReadSchema(SensorBaseSchema):
    id: int
    created_at: datetime
    updated_at: datetime
    model_config = {
        "from_attributes": True,
        "str_strip_whitespace": True,
        "str_to_lower": False,
    }
    timestamp: datetime
    value: Optional[Any] = None
    online: bool


class SensoeUpdateSchema(BaseModel):
    id: Optional[int] = None
    device_id: str
    name: Optional[str] = None
    description: Optional[str] = None
    updated_at: Optional[datetime] = None
    online: bool

    def model_post_init(self, __context):
        if not any(self.__dict__.values()):
            raise ValueError("At least one field must be provided for update")


class SensorDataReadSchema(BaseModel):
    device_id: str
    timestamp: datetime
    value: Optional[float] = None
    unit: Optional[str] = None
